### ISSUE TITLE
Feature Request: download the original .manifest file

### ISSUE THREAD

**Issue by solomax2014** [SteamRE/DepotDownloader#34]
As for now, ```-manifest-only``` argument only creates a file list. But what if someone need checksums or whatever else from .manifest (like manifest creation date)? So, if we can download manifests, can we also make the original .manifest file by calling ```-manifest-only```?

**Comment by xPaw** [comment]
What exactly do you mean by original here? Because the original would be a binary format that needs something like SteamKit to parse correctly.

I assume you want some text representation with all the attributes from the manifest? If so, how should that look?

**Comment by NicknineTheEagle** [comment]
@xPaw I think they mean saving the original .manifest from Steam CDN servers (but with decrypted filenames) like the actual Steam client does instead of using the current pseudo-format with some of the original manifest data. I would like to second this request as this feature is important for preservation purposes.

**Comment by xPaw** [comment]
> Why is a pseudo-format used anyway

DD has its own protobuf format for storing it, it is similar to Steam's, but not quite.

Dumping the raw binary manifest is pretty useless, what are you going to do with it? There aren't even any tools that support reading it directly (SteamKit has the functions to handle it obviously, but it's not directly user facing).

**Comment by NicknineTheEagle** [comment]
> DD has its own protobuf format for storing it, it is similar to Steam's, but not quite. DD using protobufs for manifests seems to predate Steam itself doing that. Steam used a different binary format before.

So this is just legacy behavior from before Steam itself was using protobuf for manifests? Why not cut out the middleman and use Steam's protobuf then?

> Dumping the raw binary manifest is pretty useless, what are you going to do with it? There aren't even any tools that support reading it directly (SteamKit has the functions to handle it obviously, but it's not directly user facing).

 You could load and parse it in SteamKit. %) As I said, this is useful for potential preservation purposes as the current Steam servers may eventually get shut down. In fact, this already happened once, Steam2 servers got shut down around 2015 so we can't get any old versions that were there anymore.

**Comment by Revadike** [comment]
A use case I can think of, is using it to integrate self-made game backups back to steam. It needs .acf files to recognize the game (which can be easily generated), but it also needs the .manifest files in order to verify/update the game backup.

**Comment by Dimensional** [comment]
_Copying from my closed requested, with a few addendums and updates._

### **What problem is this feature trying to solve?**

I'm wanting to use this program to download and save the raw CDN format of games that I have in my library. This means saving the manifest in it's raw encrypted/compressed format, and all of the 1MB CDN files that come with it. This would be used for No-Intro Datting, and possibly useful for things like Lancache, should support be enabled in the SteamKit2 library. Furthermore raw CDN dumps could be just of the files unique to the depot, and not include any common depots, like redists, so to avoid any possible duplicates, if they are not all saved to a common location.

### **How would you like it to be solved?**

Reviewing the source code, the only ways I can see possible at this time would be adding in an argument to the SteamKit2 source, specifically within Client.cs, to allow for saving the raw manifest files, and then processing it like normal to then download the raw chunks without extracting, and then adding in the necessary command-line arguments to program.cs and branch code to contentdownloader.cs. This is just preliminary, and I got no clue if there are other easier methods. Spent days going through the source code and keep finding myself going in circles or hitting dead-ends in this labyrinth.

### **Have you considered any alternative solutions**

Currently, the only viable alternate solution is the use of [SteamArchiver](https://git.sr.ht/~blowry/steamarchiver), which uses the out-dated Steam python library. It can download the CDN Chunks, the manifests, and depot keys, and can also be used to combine all the chunks into a CSD/CSM pair for that depot, which is natively supported by Steam for installing.

The only issue I see with that project right now is that due the Steam python library not being updated frequently, it doesn't support Valve's newer login mechanics, meaning that SteamArchiver can not retain any refresh token, causing logins to have to be redone over and over again. The other part is the Steam python library currently doesn't support WebSocket yet, having been requested since 2017, which means that should Valve decide to enforce that across all their CDN servers and remove TCP/UDP, every project that uses this library will stop functioning. So a viable alternative is imperative before then.

### **Additional Information**

No-Intro is a community that works to generate checksums of various games and their files for preservation. How they go about this is they set up a standard for how games should be dumped, be it a physical cart or a digital game, and how to generate a data file that contains metadata about the game files, such as what the files pertain to, their checksums, etc.

No games are ever shared, only the metadata, so that if others wish to preserve their own copies of the games, or wish to ensure any copy they have is not bootlegged, they have a means to check and confirm.

No-Intro had a tutorial for dumping game files using DepotDownloader, which has been quietly hidden away since there is a viable option to get CDN Chunks. They would prefer having the raw CDN content, ie the manifests and 1MB cdn files, depot keys, and even sidecar data from various sources including SteamDB. [Steam Preservation Proposal](https://wiki.no-intro.org/index.php?title=Steam_preservation_proposal)

**Comment by xPaw** [comment]
3.0 now saves the original Steam manifest, instead of writing DD own format.

**Comment by Dimensional** [comment]
It looks like a step closer at least, but it looks like it still processes before saving. I ran a test comparison of a manifest between the latest release and a wget download of a manifest after getting the request code for it, and the file downloaded through each have different checksums. The manifest I downloaded was the latest release of Cyberpunk 2077's main depot, https://steamdb.info/depot/1091501/history/?changeid=M:6971552143247463690.

**Comment by theraininsky** [comment]
@Dimensional no, your "wget download of a manifest" is encrypted, depotdownloader need to decrypt it before use, so it's impossible to not "processed" as you said.

**Comment by Dimensional** [comment]
> [@Dimensional](https://github.com/Dimensional) no, your "wget download of a manifest" is encrypted, depotdownloader needs to decrypt it before use, so it's impossible to not been "processed" as you said.

@theraininsky 
1. Might have missed the point of my request a small bit. To download the raw encrypted/compressed manifest and save it without processing, giving you a 1:1 copy of what is on Steam, allowing for direct examination for debugging purposes, but also because it's unmodified or processed, it can be preserved with more accuracy, keeping all of it's original data.
2. It is possible to decrypt it after it's downloaded, otherwise how would DepotDownloader be able to process the file in memory? The manifest is primarily a JSON stream that is zip compressed. The filenames are the only part that are encrypted out of the entire JSON, and that is decrypted using the depot key, which is already downloaded. https://github.com/SteamRE/SteamKit/blob/2e372027ae38c858fd48df4449ae3d3ea2c3b7d5/SteamKit2/SteamKit2/Steam/CDN/Client.cs#L70

So imagine having the file saved in it's original state, and processing it afterwards? Regardless, my request wasn't directly related to this one, which was my confusion. As this is a github issue page and not a forum thread, this discussion shouldn't shouldn't have started after a month of silence.
